### DISCUSSION TITLE
Cannot find installed software when using custom private feed

### DISCUSSION THREAD

**Discussion by trungnguyenwinget** [microsoft/winget-cli#2120] [General] [answered]
Hi all, my team are facing a problem when hosting a private feed
winget source add --name PrivateFeed **** -t Microsoft.Rest

We change package identifier and version of software in our private feed, then we install the software
winget install new_package_identifer -v new_version --source PrivateFeed

After that, we use 'list' command to find the installed software
winget list --source PrivateFeed

From the results, we cannot find the installed software because the id is different from our new_package_identifer, also the new_version.

I make this new discussion to get confirmation from winget team. Do we currently support to keep custom id and custom version from private feed?

**Comment by denelon** [comment, marked as answer]
The matching logic is based on several different factors.

We look at registry information powering Windows Apps & Features attempting to find a match with a manifest in a configured source.

If it's MSIX, we look at Package Family Name. For MSI we look at Name, Publisher, Product Code, and Upgrade Code. Similar logic is used for .exe based installers. If the manifest has the information in the "AppsAndFeatures" collection, that will improve matching further (in the future). We're still working on implementing better heuristics and matching logic to help in this area.

**Comment by denelon** [comment]
Related [Issue](https://github.com/microsoft/winget-cli/issues/2121).

**Comment by quoctienkt** [comment]
@denelon  So, you mean it depends on MSI (or MSIX) executable file, right? Is there any way to keep the package identifier and package version (force) the same as we use: winget install ... -source PrivateFeed and winget list -source PrivateFeed.
At the moment, if I change the packageIdentifier and packageVersion, it still uses information from MSI (or MSIX) file as you mentioned to show result for : winget list -source PrivateFeed.

**Comment by trungnguyenwinget** [comment]
Hi everyone, do we have any update on this? Does winget support installing software with custom packageIdentifier and packageVersion?

**Reply by denelon** [reply]
The "packageIdentifier" is a construct to have a unique identifier for a package available from a given source. Each source may have its own packages, and `winget install "packageIdentifier"` should be able to install the package if no other configured source has the same value for "packageIdentifier". If the same identifier is in more than one source, the user will need to disambiguage by specifying the source `winget install "packageIdentifier" -s <source>`.

The "packageVersion" is used to help display which version of a given package is available at a given source. If the version does not match what is reported to Windows Apps & Features, then the upgrade scenario may not work as expected. The Windows Package Manager attempts to match an installed package with the configured source based on metadata in the Windows Package Manager manifest provided by the source. We're working on an enhancement to allow packages to have a "marketing" version assigned to "packageVersion", and use the "displayVersion" in the "AppsAndFeatures" key in the Windows Package Manager Manifest as the version for comparising during the upgrade flow.
